### PR TITLE
Add project manipulator for adding references and sdks to projects

### PR DESCRIPTION
**Pull request by tautropfli**

### CODE REVIEW

File: FileManipulator/AssemblyInfo.cs [hunk 1]
```diff
@@ -0,0 +1,3 @@
+﻿using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Messerli.FileManipulator.Test")]
```
Comment (dngroth) [inline FileManipulator/AssemblyInfo.cs:3]: Add terminating newline

File: FileManipulator/Project/MsBuild/CentralPackageVersionsManipulator.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            => project.GetPropertyValue(EnableCentralPackageVersionsProperty) != FalseAsString;
+
+        private static string GetCentralPackagesFile(MsBuildProject project)
+            => project.GetPropertyValue(CentralPackagesFileProperty);
+
+        private static bool IsCentralPackageVersionsSdkImport(ResolvedImport import)
+            => import.SdkResult is { } sdk &&
+               sdk.SdkReference.Name == CentralPackageVersionsSdk;
+    }
+}
```
Comment (dngroth) [inline FileManipulator/Project/MsBuild/CentralPackageVersionsManipulator.cs:92]: Add terminating newline

File: FileManipulator/Project/MsBuild/Constant.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+    internal static class Constant
+    {
+        public const string PackageReferenceTypeTag = "PackageReference";
+        public const string VersionMetadataAttribute = "Version";
+        public const string PrivateAssetsMetadataAttribute = "PrivateAssets";
+        public const string IncludeAssetsMetadataAttribute = "IncludeAssets";
+        public const string ExcludeAssetsMetadataAttribute = "ExcludeAssets";
+        public const char ListSeparator = ';';
+    }
+}
```
Comment (dngroth) [inline FileManipulator/Project/MsBuild/Constant.cs:12]: Add terminating newline

File: FileManipulator/Project/MsBuild/ICentralPackageVersionsManipulator.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace Messerli.FileManipulator.Project.MsBuild
+{
+    internal interface ICentralPackageVersionsManipulator
+    {
+        bool ProjectUsesCentralPackageVersionsSdk(MsBuildProject project);
+
+        void AddPackageReferencesToGlobalPackages(MsBuildProject project, IEnumerable<PackageReference> packageReferences);
+    }
+}
```
Comment (dngroth) [inline FileManipulator/Project/MsBuild/ICentralPackageVersionsManipulator.cs:13]: Add terminating newline

File: FileManipulator/Project/MsBuild/IMicrosoftBuildAssemblyLoader.cs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Messerli.FileManipulator.Project.MsBuild
+{
+    internal interface IMicrosoftBuildAssemblyLoader
+    {
+        void LoadMicrosoftBuildIfNecessary();
+    }
+}
```
Comment (dngroth) [inline FileManipulator/Project/MsBuild/IMicrosoftBuildAssemblyLoader.cs:7]: Add terminating newline

File: FileManipulator/Project/MsBuild/IPackageReferenceConflictChecker.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+﻿using Messerli.FileManipulatorAbstractions.Project;
+using MsBuildProject = Microsoft.Build.Evaluation.Project;
+
+namespace Messerli.FileManipulator.Project.MsBuild
+{
+    internal interface IPackageReferenceConflictChecker
+    {
+        PackageReferenceConflictResult CheckForConflict(MsBuildProject project, PackageReference packageReference);
+    }
+}
```
Comment (dngroth) [inline FileManipulator/Project/MsBuild/IPackageReferenceConflictChecker.cs:10]: Add terminating newline

File: FileManipulator/Project/MsBuild/IProjectManipulator.cs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+﻿using Messerli.FileManipulatorAbstractions.Project;
+
+namespace Messerli.FileManipulator.Project.MsBuild
+{
+    internal interface IProjectManipulator
+    {
+        void ManipulateProject(string projectFilePath, ProjectModification modification);
+    }
+}
```
Comment (dngroth) [inline FileManipulator/Project/MsBuild/IProjectManipulator.cs:9]: Add terminating newline

File: FileManipulator/Project/MsBuild/IProjectSdkManipulator.cs [hunk 8]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+using MsBuildProject = Microsoft.Build.Evaluation.Project;
+
+namespace Messerli.FileManipulator.Project.MsBuild
+{
+    internal interface IProjectSdkManipulator
+    {
+        void AddSdksToProject(MsBuildProject project, IEnumerable<string> sdks);
+    }
+}
```
Comment (dngroth) [inline FileManipulator/Project/MsBuild/IProjectSdkManipulator.cs:10]: Add terminating newline

File: FileManipulator/Project/MsBuild/MicrosoftBuildAssemblyLoader.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    {
+        public void LoadMicrosoftBuildIfNecessary()
+        {
+            if (MSBuildLocator.CanRegister)
+            {
+                MSBuildLocator.RegisterDefaults();
+            }
+        }
+    }
+}
```
Comment (dngroth) [inline FileManipulator/Project/MsBuild/MicrosoftBuildAssemblyLoader.cs:15]: Add terminating newline

File: FileManipulator/Project/MsBuild/PackageReferenceConflictChecker.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+            return currentVersion == packageReference.Version
+                ? (PackageReferenceConflictResult)new PackageReferenceConflictResult.ExistingIsCompatible()
+                : new PackageReferenceConflictResult.ExistingIsIncompatible(currentVersion);
+        }
+
+        private static string GetVersionMetadata(ProjectItem item)
+            => item.GetMetadataValue(VersionMetadataAttribute)
+                  ?? throw new InvalidOperationException($"Existing package reference '{item.EvaluatedInclude}' is missing a version");
+    }
+}
```
Comment (dngroth) [inline FileManipulator/Project/MsBuild/PackageReferenceConflictChecker.cs:35]: Add terminating newline

File: FileManipulator/Project/MsBuild/PackageReferenceConflictResult.cs [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+        {
+            public ExistingIsIncompatible(string version)
+            {
+                Version = version;
+            }
+
+            public string Version { get; }
+        }
+    }
+}
```
Comment (dngroth) [inline FileManipulator/Project/MsBuild/PackageReferenceConflictResult.cs:23]: Add terminating newline

File: FileManipulator/Project/MsBuild/ProjectManipulator.cs [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+                DependencyAssetName.Analyzers => "analyzers",
+                DependencyAssetName.Native => "native",
+                _ => throw new InvalidOperationException($"Enum variant {dependencyAssetName} is not supported"),
+            };
+
+        private static void NoOperation<T1, T2>(T1 paramOne, T2 paramTwo)
+        {
+        }
+    }
+}
```
Comment (dngroth) [inline FileManipulator/Project/MsBuild/ProjectManipulator.cs:143]: Add terminating newline

File: FileManipulator/Project/MsBuild/CentralPackageVersionsManipulator.cs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+                .FirstOrDefault(item => item.Update == packageReference.Name);
+
+        private static string GetVersionFromMetadata(ProjectItemElement item)
+            => item.Metadata
+                .Where(m => m.Name == VersionMetadataAttribute)
+                .Select(m => m.Value)
+                .FirstOrDefault()
+                    ?? throw new InvalidOperationException($"Package reference '{item.Update}' is missing a version");
+
+        private static bool HasCentralPackageVersionsEnabled(MsBuildProject project)
```
Comment (dngroth) [inline FileManipulator/Project/MsBuild/CentralPackageVersionsManipulator.cs:82]: Would it make sense to add this function to the api?

**Review by dngroth — CHANGES_REQUESTED** [review]
Please apply `.editorconfig`

**Review by FreeApophis — APPROVED** [review]

**Pull request merged**